### ISSUE TITLE
Separate step data for each guest

### ISSUE THREAD

**Issue by psss** [teemtee/tmt#1821]
For selected steps it is necessary to create a separate workdir for each guest to prevent collision when writing data such as test results or logs. The directory structure of the `execute` step proposed on the multihost testing session looks like this:

```
plan/execute/guest/client/data
plan/execute/guest/client/results.yaml
plan/execute/guest/server/data
plan/execute/guest/server/results.yaml
plan/execute/results.yaml
```

Test logs and artifacts should be stored under `guest/<name>/data`, test results gathered from the given guest should land into `guest/<name>/results.yaml` and the test results aggregated from all guests should go into `plan/execute/results.yaml`.

In a similar way we should support separate directories for `prepare` and `finish`. I guess that for `discover`, `provision` and `report` a single directory should be enough.

**Comment by psss** [comment]
@ZelenyMartin, @lukaszachy, @happz, @sopos, @thrix, @kkaarreell, @adiosnb, any additional thoughts on this? Any volunteer to implement this? ;-)

**Comment by happz** [comment]
Ok, I forgot we have this ticket, and while going in the same direction, the prototype took a slightly different turn. So, are we safe, after the recent addition of `serialnumber` and `guest/${guest.name}/`? The answer seems to be: depends.

Plan and the corresponding output are below.

The `execute` steps seem to be fine. Every test data path now includes its serial number, plus the name of the guest it's been executed on, which means even repeated execution of the same test (name) on the same guest should lead to a different part because the repeated test (names) would have a different serial number, but, also, the name of the `discover` phase is part of the path. Compare `execute/data/guest/client-1/tests/tests/B-2` vs its repeated clone from another `discover` phase, `execute/data/guest/client-1/another-tests/tests/B-3`.

So, for the `execute`:

* single guest, test repeated in the same `discover` phase - differs by the serial number
* single guest, test repeated in multiple `discover` phases - differs by the serial number and `discover` phase name
* multiple guests, test repeated in the same `discover` phase - differs by the serial number
* multiple guests, test repeated in multiple `discover` phases - differs by the serial number, `discover` phase name, and the guest name.

WRT `results.yaml` for each `execute` incarnation: this is not implemented, there's still a single `results.yaml` with all tests. There's a `self._results: List[Result]` into which all results are added, and this is saved once `execute` is done completely. It should be possible to add this functionality, although it will require more invasive changes - yet it's aligned with what I'd like to do with `execute/internal` implementation, to avoid `self.*` use, to avoid various side-effects and setup needed before calling plugin's `go()`. But as of now, there's no conflict, we're not loosing any results.

The structure proposed above:

```
plan/execute/guest/client/data
plan/execute/guest/client/results.yaml
plan/execute/guest/server/data
plan/execute/guest/server/results.yaml
plan/execute/results.yaml
```

has become:

```
plan/execute/data/guest/client/${discover_phase.name}/tests/${test.name}-${test.serialnumber}/data
plan/execute/data/guest/server/${discover_phase.name}/tests/${test.name}-${test.serialnumber}/data
plan/execute/results.yaml
```

`prepare` and `finish` are not protected like this, there's no guest name included in a path, but: do these steps even write any data like test logs? There seems to be no need for such a sanitization because there seems to be no guest/phase-specific content saved today. I suppose that there should be a helper method, similar to `Execute.data_path()`, which would provide a path to a guest/phase-specific storage should a prepare/finish plugin need to save something. There's no such method, but there also seems to be no imminent conflict - are there files saved by `prepare` or `finish` plugins? If there are, the fix should be part of the multi-host prototype as well.

```yaml
provision:
  - how: container
    name: client-1
    role: client
    image: fedora:36

  - how: container
    name: client-2
    role: client
    image: fedora:36

  - how: container
    name: server
    role: server
    image: fedora:36

execute:
  how: tmt

prepare:
  - how: shell
    script:
      - /bin/true
    where:
      - client

  - how: shell
    script:
      - /bin/true
    where:
      - server

discover:
  - name: server-setup
    how: fmf
    test:
      - /tests/A
    where:
      - server

  - name: tests
    how: fmf
    test:
      - /tests/B
    where:
      - server
      - client

  - name: another-tests
    how: fmf
    test:
      - /tests/B
    where:
      - server
      - client

  - name: teardown
    how: fmf
    test:
      - /tests/C
```

```
/var/tmp/tmt/run-185
├── log.txt
├── plans
│   ├── data
│   ├── discover
│   │   ├── step.yaml
│   │   └── tests.yaml
│   ├── execute
│   │   ├── data
│   │   │   └── guest
│   │   │       ├── client-1
│   │   │       │   ├── another-tests
│   │   │       │   │   └── tests
│   │   │       │   │       └── B-3
│   │   │       │   │           ├── data
│   │   │       │   │           ├── metadata.yaml
│   │   │       │   │           └── output.txt
│   │   │       │   ├── teardown
│   │   │       │   │   └── tests
│   │   │       │   │       └── C-4
│   │   │       │   │           ├── data
│   │   │       │   │           ├── metadata.yaml
│   │   │       │   │           └── output.txt
│   │   │       │   └── tests
│   │   │       │       └── tests
│   │   │       │           └── B-2
│   │   │       │               ├── data
│   │   │       │               ├── metadata.yaml
│   │   │       │               └── output.txt
│   │   │       ├── client-2
│   │   │       │   ├── another-tests
│   │   │       │   │   └── tests
│   │   │       │   │       └── B-3
│   │   │       │   │           ├── data
│   │   │       │   │           ├── metadata.yaml
│   │   │       │   │           └── output.txt
│   │   │       │   ├── teardown
│   │   │       │   │   └── tests
│   │   │       │   │       └── C-4
│   │   │       │   │           ├── data
│   │   │       │   │           ├── metadata.yaml
│   │   │       │   │           └── output.txt
│   │   │       │   └── tests
│   │   │       │       └── tests
│   │   │       │           └── B-2
│   │   │       │               ├── data
│   │   │       │               ├── metadata.yaml
│   │   │       │               └── output.txt
│   │   │       └── server
│   │   │           ├── another-tests
│   │   │           │   └── tests
│   │   │           │       └── B-3
│   │   │           │           ├── data
│   │   │           │           ├── metadata.yaml
│   │   │           │           └── output.txt
│   │   │           ├── server-setup
│   │   │           │   └── tests
│   │   │           │       └── A-1
│   │   │           │           ├── data
│   │   │           │           ├── metadata.yaml
│   │   │           │           └── output.txt
│   │   │           ├── teardown
│   │   │           │   └── tests
│   │   │           │       └── C-4
│   │   │           │           ├── data
│   │   │           │           ├── metadata.yaml
│   │   │           │           └── output.txt
│   │   │           └── tests
│   │   │               └── tests
│   │   │                   └── B-2
│   │   │                       ├── data
│   │   │                       ├── metadata.yaml
│   │   │                       └── output.txt
│   │   ├── results.yaml
│   │   └── step.yaml
│   ├── finish
│   │   └── step.yaml
│   ├── prepare
│   │   └── step.yaml
│   ├── provision
│   │   ├── guests.yaml
│   │   └── step.yaml
│   └── report
│       └── step.yaml
└── run.yaml

54 directories, 31 files
```

**Comment by thrix** [comment]
@happz can we also have here an example how this will affect the `results.yaml`?

**Comment by happz** [comment]
@thrix absolutely, here you go. Note the added `serialnumber` and `guest` and their appearance in test names and log paths:

```yaml
  - name: /server-setup/tests/A
    serialnumber: 1
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-1/server-setup/tests/A-1/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /tests/tests/B
    serialnumber: 2
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-1/tests/tests/B-2/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /another-tests/tests/B
    serialnumber: 3
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-1/another-tests/tests/B-3/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /teardown/tests/C
    serialnumber: 4
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-1/teardown/tests/C-4/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /server-setup/tests/A
    serialnumber: 1
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-2/server-setup/tests/A-1/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /tests/tests/B
    serialnumber: 2
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-2/tests/tests/B-2/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /another-tests/tests/B
    serialnumber: 3
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-2/another-tests/tests/B-3/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /teardown/tests/C
    serialnumber: 4
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/client-2/teardown/tests/C-4/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /server-setup/tests/A
    serialnumber: 1
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/server/server-setup/tests/A-1/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /tests/tests/B
    serialnumber: 2
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/server/tests/tests/B-2/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /another-tests/tests/B
    serialnumber: 3
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/server/another-tests/tests/B-3/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
  - name: /teardown/tests/C
    serialnumber: 4
    result: pass
    note:
    duration: 00:00:00
    ids:
        id:
        extra-nitrate:
        extra-task:
    log:
      - data/guest/server/teardown/tests/C-4/output.txt
    guest:
        name:
        role:
        __class__:
            module: tmt.result
            name: ResultGuestData
    __class__:
        module: tmt.result
        name: Result
```

**Comment by psss** [comment]
Thanks for providing an excellent overview and analysis. I believe the essential part of this issue has been covered. If we need to handle any possible collisions in the `prepare` and `finish` steps let's handle that in a separate issue later, but as for now I don't see any imminent danger there.
